src/coprocessor/read_pool_impl.rs

```diff
@@ -0,0 +1,144 @@
+        RefCell::new(COPR_SCAN_KEYS.local());
+    pub static LOCAL_COPR_SCAN_DETAILS: RefCell<LocalIntCounterVec> =
+        RefCell::new(COPR_SCAN_DETAILS.local());
+    pub static LOCAL_COPR_ROCKSDB_PERF_COUNTER: RefCell<LocalIntCounterVec> =
+        RefCell::new(COPR_ROCKSDB_PERF_COUNTER.local());
+    static LOCAL_COPR_EXECUTOR_COUNT: RefCell<LocalIntCounterVec> =
+        RefCell::new(COPR_EXECUTOR_COUNT.local());
+    static LOCAL_COPR_GET_OR_SCAN_COUNT: RefCell<LocalIntCounterVec> =
+        RefCell::new(COPR_GET_OR_SCAN_COUNT.local());
+
+    static LOCAL_PD_SENDER: RefCell<Option<FutureScheduler<PdTask>>> =
+        RefCell::new(None);
+
+    static LOCAL_COP_FLOW_STATS: RefCell<HashMap<u64, crate::storage::FlowStatistics>> =
+        RefCell::new(HashMap::default());
+}
+
+pub struct ReadPoolImpl;
+
+impl std::fmt::Debug for ReadPoolImpl {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        fmt.debug_struct("coprocessor::ReadPoolImpl").finish()
+    }
+}
+
+impl ReadPoolImpl {
+    pub fn build_read_pool(
+        config: &readpool::Config,
+        pd_sender: FutureScheduler<PdTask>,
+        name_prefix: &str,
+    ) -> ReadPool {
+        let pd_sender2 = pd_sender.clone();
+
+        Builder::from_config(config)
+            .name_prefix(name_prefix)
+            .on_tick(move || ReadPoolImpl::thread_local_flush(&pd_sender))
+            .before_stop(move || ReadPoolImpl::thread_local_flush(&pd_sender2))
+            .build()
+    }
+
+    #[inline]
+    fn thread_local_flush(pd_sender: &FutureScheduler<PdTask>) {
+        // Flush Prometheus metrics
+        LOCAL_COPR_REQ_HISTOGRAM_VEC.with(|m| m.borrow_mut().flush());
+        LOCAL_OUTDATED_REQ_WAIT_TIME.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_REQ_HANDLE_TIME.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_REQ_WAIT_TIME.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_REQ_ERROR.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_SCAN_KEYS.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_ROCKSDB_PERF_COUNTER.with(|m| m.borrow_mut().flush());
+
+        // Report PD metrics
+        LOCAL_COP_FLOW_STATS.with(|local_cop_flow_stats| {
+            if local_cop_flow_stats.borrow().is_empty() {
+                // Stats to report to PD is empty, ignore.
+                return;
+            }
+
+            let read_stats = local_cop_flow_stats.replace(HashMap::default());
+            let result = pd_sender.schedule(PdTask::ReadStats { read_stats });
+            if let Err(e) = result {
+                error!("Failed to send cop pool read flow statistics"; "err" => ?e);
+            }
+        });
+
+        LOCAL_COPR_SCAN_DETAILS.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_GET_OR_SCAN_COUNT.with(|m| m.borrow_mut().flush());
+        LOCAL_COPR_EXECUTOR_COUNT.with(|m| m.borrow_mut().flush());
+    }
+
+    pub fn collect(region_id: u64, type_str: &str, metrics: ExecutorMetrics) {
```

Comment:
let's rename it to make it more clear. maybe.. `thread_local_collect_executor_metrics`?